src/tree/split_evaluator.cc

```diff
@@ -0,0 +1,251 @@
+    bst_float ret = 0.0f;
+
+    for(auto& e : m_inner) {
+      ret += e->ComputeWeight(parentID, stats);
+    }
+
+    return ret;
+  }
+
+ private:
+  std::vector<std::unique_ptr<SplitEvaluator> > m_inner;
+};
+
+SplitEvaluator* SplitEvaluator::Create(const std::string& name) {
+
+  std::vector<std::string> names = common::Split(name, ',');
+
+  /* If more than one split evaluator was specified (e.g., RidgePenalty and MonotonicConstraint), merge them using
+     AdditiveSplitEvaluator. */
+  if(names.size() > 1) {
+    std::vector<std::unique_ptr<SplitEvaluator> > inner;
+
+    for(auto& n : names) {
+      auto* e = ::dmlc::Registry< ::xgboost::tree::SplitEvaluatorReg>::Get()->Find(name);
+      if (e == nullptr) {
+        LOG(FATAL) << "Unknown SplitEvaluator " << name;
+      }
+      inner.push_back(std::move(std::unique_ptr<SplitEvaluator>((e->body)())));
+    }
+
+    return new AdditiveSplitEvaluator(inner);
+  }
+  else {
+    auto* e = ::dmlc::Registry< ::xgboost::tree::SplitEvaluatorReg>::Get()->Find(name);
+    if (e == nullptr) {
+      LOG(FATAL) << "Unknown SplitEvaluator " << name;
+    }
+    return (e->body)();
+  }
+}
+
+// Default implementations of some virtual methods that that need to do anything by default
+SplitEvaluator::~SplitEvaluator() {}
+void SplitEvaluator::Init(const std::vector<std::pair<std::string, std::string> >& args) {}
+void SplitEvaluator::Reset() {}
+
+bst_float SplitEvaluator::ComputeSplitLoss(bst_uint parentID,
+                                           bst_uint featureID,
+                                           const GradStats& left,
+                                           const GradStats& right) const {
+  return ComputeLoss(parentID, left) + ComputeLoss(parentID, right);
+}
+
+/*! \brief Encapsulates the parameters required by the RidgePenalty split evaluator */
+struct RidgePenaltyParams : public dmlc::Parameter<RidgePenaltyParams> {
+  float reg_lambda;
+  float reg_gamma;
+
+  DMLC_DECLARE_PARAMETER(RidgePenaltyParams) {
+    DMLC_DECLARE_FIELD(reg_lambda)
+      .set_lower_bound(0.0)
+      .set_default(1.0)
+      .describe("L2 regularization on leaf weight");
+    DMLC_DECLARE_FIELD(reg_gamma)
+      .set_lower_bound(0.0f)
+      .set_default(0.0f)
+      .describe("Cost incurred by adding a new leaf node to the tree");
+    DMLC_DECLARE_ALIAS(reg_lambda, lambda);
+    DMLC_DECLARE_ALIAS(reg_gamma, gamma);
+  }
+};
+
+DMLC_REGISTER_PARAMETER(RidgePenaltyParams);
+
+/*! \brief Applies an L2 penalty and per-leaf penalty. */
+class RidgePenalty : public SplitEvaluator {
+ public:
+  void Init(const std::vector<std::pair<std::string, std::string> >& args) override {
+    m_params.InitAllowUnknown(args);
+  }
+
+  SplitEvaluator* GetHostClone() const override {
+    RidgePenalty* r = new RidgePenalty();
+    r->m_params = this->m_params;
+
+    return r;
+  }
+
+  bst_float ComputeLoss(bst_uint parentID, const GradStats& stats) const override {
```

Comment:
Is this a loss function or a gain function?

Reply:
This should be the loss function.